File: app/src/main/java/ru/darkcat/camera/upload/UploadWorker.java
package ru.darkcat.camera.upload;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import java.io.File;

import ru.darkcat.camera.data.DarkCatDatabase;
import ru.darkcat.camera.data.DarkCatSettings;
import ru.darkcat.camera.data.MediaRecord;
import ru.darkcat.camera.vault.VaultRepository;

/** Executes encrypted-media upload without ever gating capture. */
public final class UploadWorker extends Worker {
    public UploadWorker(@NonNull Context context, @NonNull WorkerParameters parameters) {
        super(context, parameters);
    }

    @NonNull @Override public Result doWork() {
        SyncDiagnostics.recordStart(getApplicationContext());
        String id = getInputData().getString("media_id");
        if (id == null) return Result.failure();
        DarkCatDatabase database = DarkCatDatabase.get(getApplicationContext());
        MediaRecord record = database.get(id);
        if (record == null) return Result.failure();

        if (record.status == MediaRecord.UploadStatus.LOCAL_DELETE_PENDING) {
            return resumeLocalDeletion(database, record);
        }
        if (record.status == MediaRecord.UploadStatus.VERIFIED
                || record.status == MediaRecord.UploadStatus.LOCAL_DELETED) { SyncDiagnostics.recordSuccess(getApplicationContext()); return Result.success(); }
        if (!UploadStateMachine.canStartUpload(record.status)) return Result.failure();

        int attempt = record.retryCount + 1;
        try {
            record = database.transitionStatus(id, MediaRecord.UploadStatus.UPLOADING, null, attempt);
            File file = new VaultRepository(getApplicationContext()).vaultFile(record.vaultFileName);
            if (!file.isFile() || file.length() != record.encryptedSize) {
                return recordFailure(database, id, attempt, "Encrypted vault file is missing or has changed");
            }
            UploadProvider.UploadResult upload = UploadProviders.forContext(getApplicationContext())
                    .upload(getApplicationContext(), record, file);
            if (!upload.accepted) return recordFailure(database, id, attempt, upload.message);

            record = database.transitionStatus(id, MediaRecord.UploadStatus.UPLOADED, upload.message, attempt);
            DarkCatSettings.set(getApplicationContext(), "darkcat_last_sync_success",
                    System.currentTimeMillis());
            if (!upload.verified) { SyncDiagnostics.recordSuccess(getApplicationContext()); return Result.success(); }

            record = database.transitionStatus(id, MediaRecord.UploadStatus.VERIFIED, upload.message, attempt);
            if (!UploadStateMachine.shouldDeleteLocal(record.status,
                    DarkCatSettings.deleteAfterVerified(getApplicationContext()))) { SyncDiagnostics.recordSuccess(getApplicationContext()); return Result.success(); }
            record = database.transitionStatus(id, MediaRecord.UploadStatus.LOCAL_DELETE_PENDING, null, attempt);
            return resumeLocalDeletion(database, record);
        } catch (Exception error) {
            SyncDiagnostics.recordFailure(getApplicationContext(), attempt, safeError(error));
            MediaRecord latest = database.get(id);
            if (latest != null && latest.status == MediaRecord.UploadStatus.LOCAL_DELETE_PENDING) {
                try { database.transitionStatus(id, latest.status, safeError(error), latest.retryCount); }
                catch (Exception ignored) { }
                return Result.retry();
            }
            if (latest != null && (latest.status == MediaRecord.UploadStatus.UPLOADING
                    || latest.status == MediaRecord.UploadStatus.QUEUED
                    || latest.status == MediaRecord.UploadStatus.FAILED_RETRYABLE)) {
                return recordFailure(database, id, attempt, safeError(error));
            }
            // UPLOADED is intentionally retained if only verification/state persistence failed.
            return Result.retry();
        }
    }

    private Result resumeLocalDeletion(DarkCatDatabase database, MediaRecord record) {
        try {
            new VaultRepository(getApplicationContext()).deleteFilesChecked(record);
            database.transitionStatus(record.id, MediaRecord.UploadStatus.LOCAL_DELETED, null, record.retryCount);
            return Result.success();
        } catch (Exception error) {
            try { database.transitionStatus(record.id, MediaRecord.UploadStatus.LOCAL_DELETE_PENDING,
                    safeError(error), record.retryCount); }
            catch (Exception ignored) { }
            return Result.retry();
        }
    }

    private Result recordFailure(DarkCatDatabase database, String id, int attempt, String message) {
        SyncDiagnostics.recordFailure(getApplicationContext(), attempt, message);
        MediaRecord.UploadStatus failure = UploadStateMachine.failureStatus(attempt);
        try { database.transitionStatus(id, failure, message, attempt); }
        catch (Exception ignored) { return Result.retry(); }
        return failure == MediaRecord.UploadStatus.FAILED_PERMANENT ? Result.failure() : Result.retry();
    }

    private static String safeError(Exception error) {
        return error == null ? "Unknown upload error" : error.getClass().getSimpleName();
    }
}

File: app/src/main/java/ru/darkcat/camera/upload/UploadWorker.kt
package ru.darkcat.camera.upload

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import ru.darkcat.camera.DarkCatApplication
import ru.darkcat.camera.data.MediaRecord
import ru.darkcat.camera.data.UploadStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class UploadWorker(appContext: Context, params: WorkerParameters) : CoroutineWorker(appContext, params) {
    override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
        val mediaId = inputData.getString(MEDIA_ID_KEY) ?: return@withContext Result.failure()
        val repository = (applicationContext as DarkCatApplication).vaultRepository
        val record = repository.get(mediaId) ?: return@withContext Result.failure()
        if (record.uploadStatus == UploadStatus.VERIFIED || record.uploadStatus == UploadStatus.LOCAL_DELETED) {
            return@withContext Result.success()
        }
        val currentRetryCount = record.retryCount + 1
        repository.updateUploadStatus(mediaId, UploadStatus.UPLOADING, retryCount = currentRetryCount)
        try {
            val receipt = MediaUploaderRegistry.factory(applicationContext).upload(
                UploadRequest(
                    mediaId = mediaId,
                    idempotencyKey = mediaId,
                    encryptedFile = repository.encryptedFile(record),
                    checksumSha256 = record.checksumSha256,
                    mimeType = record.mimeType,
                ),
            )
            if (receipt.checksumSha256 != record.checksumSha256) {
                throw RetryableUploadException("Server checksum confirmation did not match local checksum")
            }
            repository.updateUploadStatus(mediaId, UploadStatus.UPLOADED, uploadId = receipt.uploadId)
            repository.updateUploadStatus(mediaId, UploadStatus.VERIFIED, uploadId = receipt.uploadId)
            Result.success()
        } catch (error: RetryableUploadException) {
            repository.updateUploadStatus(mediaId, UploadStatus.FAILED_RETRYABLE, error = error.message, retryCount = currentRetryCount)
            Result.retry()
        } catch (error: PermanentUploadException) {
            repository.updateUploadStatus(mediaId, UploadStatus.FAILED_PERMANENT, error = error.message, retryCount = currentRetryCount)
            Result.failure()
        } catch (error: Throwable) {
            repository.updateUploadStatus(mediaId, UploadStatus.FAILED_RETRYABLE, error = error.message, retryCount = currentRetryCount)
            Result.retry()
        }
    }

    companion object {
        const val MEDIA_ID_KEY = "media_id"
    }
}
